### DISCUSSION TITLE
How to disable type signatures in instances?

### DISCUSSION THREAD

**Discussion by hellwolf** [haskell/haskell-language-server#3105] [General]
After a recent hls update, the type signatures in the instances are cluttering my Emacs editor:

![image](https://user-images.githubusercontent.com/186660/184629202-3f7ac771-3c81-41fa-8cfc-d1259cc1435a.png)

I often accidentally click them and it would add instance sig for me, I'd prefer not seeing them.

Thank you in advance!

**Comment by jhrcek** [comment]
Look for a way to disable "class plugin" in your editor.
Sorry for answering question you didn't ask, but here's how this can be done in vscode:

https://user-images.githubusercontent.com/2716069/184631617-976c7916-a8e8-41ef-9e34-e5d1d9f3863e.mp4

**Comment by July541** [comment]
Hello @hellwolf, sorry for bothering you.

You can disable it by `"haskell.plugin.class.codeLensOn": false` in vscode, not sure about other editors. Run `haskell-language-server generate-default-config` may have some hints.

BTW, you can query all settings by `haskell-language-server vscode-extension-schema`, and further information can be found at https://haskell-language-server.readthedocs.io/en/latest/configuration.html

**Comment by hellwolf** [comment]
Thank you @jhrcek and @July541 ! 

Since I am using emacs, I am trying to figure out a way myself:

`lsp-haskell-plugin-class-global-on` definitely worked for me, but trying to disable codeLensOn failed so far, I will dig a bit more details later. Maybe some emacs user could see this message and comment what I am doing wrong:

(EDIT: WRONG CODE)
```elisp
(use-package lsp-haskell
  :config
  (add-hook 'lsp-after-initialize-hook
          (lambda ()
            (lsp--set-configuration
              '(:haskell (:plugin (:class (:codeLensOn false))))
              ))))
```

**Comment by hellwolf** [comment]
Just wanted to say that I like the feature, just that the clutterness and the fact clicking it accidentally sometimes modifies the code is a bit undesired.

Thank you for all your works, I enjoy it a lot using emacs.

**Comment by July541** [comment]
@hellwolf From https://github.com/emacs-lsp/lsp-haskell#configuration, maybe 
```
(use-package lsp-haskell
  :config
  (add-hook 'lsp-after-initialize-hook
          '(lambda () ;; note here
            (lsp--set-configuration
              '(:haskell (:plugin (:class (:codeLensOn false))))
              ))))

```

Just from an amateur of emacs.

**Comment by hellwolf** [comment]
![image](https://user-images.githubusercontent.com/186660/184647013-62509e33-8fce-446f-a72f-079f632d6f38.png)

that one definitely helped, but now I need to fight for the "false" being string issue

**Comment by hellwolf** [comment]
I think I found the right combination:

```elisp
(use-package lsp-haskell
  :config
  (add-hook 'lsp-after-initialize-hook
            #'(lambda ()
                (lsp--set-configuration
                 '(:haskell (:plugin (:class (:codeLensOn nil))))
                 ))))
```

:)

EDIT: still didn't work

**Comment by hellwolf** [comment]
I will report to lsp-mode my experience, maybe they could make it different. I am not sure how vscode experience is.

**Comment by July541** [comment]
> I am not sure how vscode experience is.

I am sure about adding `"haskell.plugin.class.codeLensOn": false` to config file is enough :)

**Reply by hellwolf** [reply]
Oh, I meant the plugin experience itself. I look for disabling it because I often miscliked instance sig and it adds to the source code, which I really didn't want to.

**Reply by hellwolf** [reply]
I'd not disable it if it didn't do that.

**Reply by hellwolf** [reply]
Btw, I am using "master" branch of the hls compiled last week.

**Reply by July541** [reply]
I can't find a better solution in a short term, someone who has ideas are welcome :)

**Reply by July541** [reply]
Your lenses are cluttered indeed, they need a better UI like it in vscode.

**Reply by hellwolf** [reply]
Do you not accidentally click those type signatures? What happens in vscode if you clicked them?

**Reply by July541** [reply]
vscode will also insert a signature. I seldom click it by mistake and I use the keyboard to navigate the code.

**Comment by michaelpj** [comment]
I'm surprised that the lenses are appearing to the *right* of the definition, rather than above them like they do for other type signature lenses. Is that intended @July541 ?

**Reply by michaelpj** [reply]
Hmm, in @jhrcek 's video of vscode they're appearing above. Something is strange there. Maybe emacs interprets the position differently? Are we sending something weird?

**Reply by July541** [reply]
> I'm surprised that the lenses are appearing to the right of the definition, rather than above them like they do for other type signature lenses. 

Me too. I just thought it is how emacs implement code lens, look not?

They should be displayed above the code. But how to render the lens relies on editor's implementation.

**Comment by michaelpj** [comment]
@hellwolf could you give us the logs of a session showing the type lens responses. And can you see if you get the same behaviour as @jhrcek in vscode?

**Reply by hellwolf** [reply]
https://pastebin.com/afFTQTfD

**Reply by hellwolf** [reply]
@michaelpj let me know if this helps

**Reply by July541** [reply]
It looks like little info is available, I'd love to see the log from emacs session while rendering lsp requests.

Indeed, we should log the position info for lens, but it doesn't help for editor's difference...

**Comment by michaelpj** [comment]
Okay, now I can reproduce this!

**Reply by michaelpj** [reply]
I'll try and get a session log and make a proper issue.

**Reply by hellwolf** [reply]
Looking forward to, definitely want to learn also of how to be a contributor, using this opportunity...

**Reply by michaelpj** [reply]
All the code lenses were doing this, so I had a look and I found this in the `lsp-mode` [changelog](https://github.com/emacs-lsp/lsp-mode/blob/master/CHANGELOG.org#release-800):

> Add lsp-lens-place-position with option to place code lens at end of line as default.

So that explains the placement to the right, which was mysterious. As for turning them off, I think that's answered already.

**Comment by hellwolf** [comment]
Yes

```
  (lsp-lens-place-position 'above-line)
```
![image](https://user-images.githubusercontent.com/186660/194316637-4a830c5d-0589-42e7-ba94-fcd818d3bdc0.png)

Works!

Thanks @michaelpj .

I should look into changelog first next!

**Comment by hellwolf** [comment]
Actually the real annoyance is still that I sometimes accidentally clock into those "code lenses" then it creates those code for me...

**Reply by michaelpj** [reply]
Sure, but you can turn them off.

**Reply by hellwolf** [reply]
oh, I have been trying to find how, could you please hint where can I find the option for that?

**Reply by July541** [reply]
Unfortunately, you can only disable the lens instead of disabling the clicking action...

But we can replace the lens with inlay hints in the future. https://github.com/haskell/haskell-language-server/issues/2938#issuecomment-1270175897
